### PR TITLE
Show better diagnostics for unicode escapes

### PR DESCRIPTION
**Pull request by nagisa**
Inspired by the now-mysteriously-closed https://github.com/rust-lang/rust/pull/26782.

This PR introduces better error messages when unicode escapes have invalid format (e.g. `\uFFFF`). It also makes rustc always tell the user that escape may not be used in byte-strings and bytes and fixes some spans to not include unecessary characters and include escape backslash in some others.

### CODE REVIEW

File: src/libsyntax/parse/lexer/mod.rs [hunk 1]
```diff
@@ -728,19 +738,25 @@ impl<'a> StringReader<'a> {
...
-                            'u' if self.curr_is('{') => {
-                                let valid = self.scan_unicode_escape(delim);
-                                if valid && ascii_only {
-                                    self.err_span_(
-                                        escaped_pos,
-                                        self.last_pos,
+                            'u' => {
+                                let valid = if self.curr_is('{') {
+                                    let valid = self.scan_unicode_escape(delim);
+                                    if ascii_only { false } else { valid }
```
Comment (nrc) [inline src/libsyntax/parse/lexer/mod.rs]: `!ascii_only && self.scan_unicode_escape(delim)`

File: src/libsyntax/parse/lexer/mod.rs [hunk 2]
```diff
@@ -728,19 +738,25 @@ impl<'a> StringReader<'a> {
...
+                                } else {
+                                    self.err_span_(start, self.last_pos,
+                                        "incomplete unicode escape sequence");
```
Comment (nrc) [inline src/libsyntax/parse/lexer/mod.rs]: s/incomplete/incorrect

**Comment by nrc** [comment]
r+ with the nits addressed

**Comment by nagisa** [comment]
Comments addressed.

**Pull request merged**